### ISSUE TITLE
Site rendering is much slower with taxonomies

### ISSUE THREAD

**Issue by scouten** [Keats/gutenberg#1897]
# Bug Report

## Environment

Zola version: 0.15.3

## Expected Behavior
Site generation remains speedy.

## Current Behavior
I have a (closed source, sorry) site with about 800 pages, 120 sections, and two taxonomies. There is at least one entry per taxonomy per page.

When I added the taxonomies, the site rendering time dropped from about 1.5 seconds to ~20 seconds on a very beefy M1 MacBook Pro laptop. Running with `zola serve` often fails to reload properly.

## Step to reproduce
If it would be useful, I could take the time to build an obfuscated version of the site. It would take me some hours, so please don't ask lightly.

```sh
$ zola build
Building site...
Checking all internal links with anchors.
> Successfully checked 0 internal link(s) with anchors.
-> Creating 791 pages (0 orphan) and 124 sections
Done in 20.9s.
```

**Comment by Keats** [comment]
Can you try with the `next` branch?

**Comment by scouten** [comment]
No appreciable difference. Same site just rendered in 20.3 seconds using the latest `next` branch.

**Comment by Keats** [comment]
Hm I think I would need the obfuscated version :/ Is it not possible to just replace the content in .md files after the front-matter with lorem ipsum? That should be pretty easy with a small script and you can send the archive to my email directly so it's not public.
I can write a script for that if you want too

**Comment by Keats** [comment]
Something like

```python
import glob
import sys
import shutil

NEW_TEXT = """

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Phasellus quis dui tellus. Vestibulum consequat arcu eu porta egestas. Donec urna ligula, convallis sit amet convallis sed, ultrices vel ligula. Integer congue vitae felis ut ultricies. Duis a dolor ornare lectus fermentum laoreet condimentum vitae est. Praesent arcu nibh, pulvinar in congue at, semper sed arcu. Aliquam erat volutpat. Cras at sollicitudin libero, quis vestibulum est. Phasellus sagittis bibendum metus, sit amet laoreet nibh auctor eu. Maecenas mauris urna, auctor sed volutpat tempor, fermentum ut nisi. Morbi ut dolor nisl. Fusce pulvinar sem id rhoncus aliquet. Etiam pellentesque condimentum blandit. 
"""

source = sys.argv[1]
destination = sys.argv[2]

try:
    shutil.rmtree(destination)
except:
    pass

shutil.copytree(source, destination)
# So we can't touch it anymore
source = None

paths = glob.glob(f"{destination}/content/**/*.md", recursive=True)

for p in paths:
    with open(p, "r+") as f:
        print(f"Opened {p}")
        content = f.read()
        f.truncate(0)
        f.seek(0)
        new_content = ""
        fm_markers_seen = 0
        for l in content.splitlines():
            if l.startswith("+++"):
                fm_markers_seen += 1
            new_content += l
            new_content += "\n"
            if fm_markers_seen == 2:
                break
        new_content += NEW_TEXT
        f.write(new_content)
```

should work. You can also add title anonymisation if it's important

To use `python script.py THE_SITE_PATH NEW_PATH`. Be careful as it will erase `NEW_PATH` first

**Comment by Keats** [comment]
ping @scouten i'd like to fix that before releasing 0.16

**Comment by scouten** [comment]
@Keats sorry, been extra busy with [other Rust-related work](https://opensource.contentauthenticity.org) last few weeks. Catching up and I started on building the anonymized version of my site last night. Hope to have something to share in next day or two.

**Comment by scouten** [comment]
Thanks for the pointer on replacing the text. I was able to create a repro case which shows the problem pretty clearly: https://github.com/scouten/special-waffle

tl;dr: By removing the two taxonomies in the anonymized site, rendering time dropped from 16.0 seconds to 2.8 seconds on my Mac M1 MAX laptop.

**Comment by Keats** [comment]
> been extra busy with [other Rust-related work](https://opensource.contentauthenticity.org/) last few weeks.

Looks sweet, didn't realise Adobe was using Rust!

> I was able to create a repro case which shows the problem pretty clearly: https://github.com/scouten/special-waffle

You need to give me access to the repo but that sounds perfect!

**Comment by scouten** [comment]
> > been extra busy with [other Rust-related work](https://opensource.contentauthenticity.org/) last few weeks.
> 
> Looks sweet, didn't realise Adobe was using Rust!

My team is the most public about it, but there's quite a bit of interest brewing.

> > I was able to create a repro case which shows the problem pretty clearly: https://github.com/scouten/special-waffle
> 
> You need to give me access to the repo but that sounds perfect!

Just sent you a collaborator invite.

**Comment by scouten** [comment]
@Keats thank you for the diagnosis on my site. If you look at how I'm using `get_taxonomy`\, what I really wanted was the one tag and the count of pages for that tag. Would you be open to a PR that added a get term by name built-in function, possibly with an "omit pages" argument? That would at least sharply reduce the volume of data being serialized without radically changing the architecture.

**Comment by Keats** [comment]
Probably not for the next release but maybe for the one after.
